File: LeetCode/CPP/LeetCode90.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

void recurseAndBacktrack(int idx, vector<vector<int>>& allsets, vector<int>& currentset, vector<int>& nums){
    allsets.push_back(currentset);
    for(int i=idx;i<nums.size();i++){
        if(i>idx && nums[i]==nums[i-1]){
            continue;
        }
        currentset.push_back(nums[i]);
        recurseAndBacktrack(i+1, allsets, currentset, nums);
        currentset.pop_back();
    }
}

vector<vector<int>> findSubsets(vector<int>& nums){
    sort(nums.begin(), nums.end());
    vector<vector<int>> allsets;
    vector<int> currentSet;
    recurseAndBacktrack(0, allsets, currentSet, nums);
    return allsets;
}

int main(){
    vector<int> nums = {1, 2, 2};
    vector<vector<int>> subsets = findSubsets(nums);
    for(vector<int> subset : subsets){
        cout<<"[ ";
        for(int num : subset){
            cout<<num<<" ";
        }
        cout<<"]"<<endl;
    }
    return 0;
}

File: LeetCode/JAVA/LeetCode90.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class LeetCode90 {

    public static void recurseAndBacktrack(int idx, List<List<Integer>> allsets, List<Integer> currset, int[] nums){
        allsets.add(new ArrayList<>(currset));
        for(int i=idx;i<nums.length;i++){
            if(i>idx && nums[i]==nums[i-1]){
                continue;
            }
            currset.add(nums[i]);
            recurseAndBacktrack(i+1, allsets, currset, nums);
            currset.remove(currset.size()-1);
        }
    }

    public static List<List<Integer>> findSubsets(int[] nums){
        Arrays.sort(nums);
        List<List<Integer>> allsets = new ArrayList<>();
        List<Integer> currset = new ArrayList<>();
        recurseAndBacktrack(0, allsets, currset, nums);
        return allsets;
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 2};
        List<List<Integer>> res = findSubsets(nums);
        System.out.println(res);
    }
}
